src/libcore/raw.rs

```diff
@@ -97,3 +97,30 @@ pub struct TraitObject {
     pub data: *mut (),
     pub vtable: *mut (),
 }
+
+/// An adapter from a generator to a future.
+///
+/// This is a permanently unstable API used by the compiler. You should not
+/// depend on it.
+#[allow(missing_debug_implementations, dead_code)]
+#[unstable(feature = "gen_future", issue = "50547")]
+pub struct GenFuture<G, R> {
```

Comment:
If you add a bound on `G` that mentions `R`, I think you can both get rid of the `PhantomData`, *and* force the type-checker to infer the return type from the type you put in `R`, from `GenFuture` alone, although `gen_future` below is probably enough either way.

Reply:
I didn't know you could get rid of PhantomData this way. If that works I'll just put the bound here, make the field public, and get rid of the constructor.